### PR TITLE
Intrusive elem and cond pointers

### PR DESCRIPTION
**Pull request by RiccardoRossi**
this PR makes pointers to elements and conditions to be intrusive

### CODE REVIEW

**Comment by loumalouomega** [comment]
I will try as soon as the other fix PR are merged

File: kratos/includes/element.h [hunk 1]
```diff
@@ -63,7 +63,8 @@ class Element : public GeometricalObject, public Flags
     ///@name Type Definitions
     ///@{
     /// Pointer definition of Element
-    KRATOS_CLASS_POINTER_DEFINITION(Element);
+    typedef Kratos::intrusive_ptr<Element> Pointer;
+    typedef Kratos::intrusive_weak_ptr<Element> WeakPointer;
```
Comment (philbucher) [inline kratos/includes/element.h]: why not `KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION`?

File: kratos/includes/condition.h [hunk 2]
```diff
@@ -63,7 +63,9 @@ class Condition : public GeometricalObject, public Flags
     ///@{
 
     /// Pointer definition of Condition
-    KRATOS_CLASS_POINTER_DEFINITION(Condition);
+    typedef Kratos::intrusive_ptr<Condition> Pointer;
+    typedef Kratos::intrusive_weak_ptr<Condition> WeakPointer;
+    typedef Kratos::unique_ptr<Condition> UniquePointer;
```
Comment (philbucher) [inline kratos/includes/condition.h]: same

Comment (RiccardoRossi) [reply kratos/includes/condition.h]: the only reason is that i added it manually before adding the macro.

i ll change this tomorrow

File: kratos/includes/element.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/includes/element.h]: Indeed

**Comment by loumalouomega** [comment]
It does compile from my side

**Comment by RiccardoRossi** [comment]
just to report that this also compiles and runs the DEM stuff (according to @maceligueta , please correct me if i am wrong).

@philbucher could u check if this is also ok for your stuff? 

@rubenzorrilla ?

if possible i would like to merge this soon (guess that beginning of next week) to be followed right after from #4729 which depends on it. This means that if you also have the time to verify that #4792 compiles and report it in that branch i would be grateful.

**Comment by philbucher** [comment]
this branch works for me (ParticleMechanics is not compiling FYI @bodhinandach @antonialarese )

The neighbors as GP I will test once you updated it

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/altair please note this change. It would be great if you can check it before merging to master.

**Comment by ddiezrod** [comment]
@KratosMultiphysics/altair @pooyan-dadvand  @RiccardoRossi I checked and all changes needed are trivial from our side.  Only error is when doing:
```
WeakPointerVector<Element> fathers;
Element::Pointer pcurrent_element=Element::Pointer(fathers(0)); 
```
How should be this now?

**Review by pooyan-dadvand — DISMISSED** [review]
:+1:

**Comment by RiccardoRossi** [comment]
@ddiezrod if you need to USE the pointer, without creating explicitly a new shared pointer you need to do:

~~~cpp
GlobalPointersVector<Element> fathers;
auto pelem = fathers[0]; 
KRATOS_WATCH(pelem->id());
~~~

if you actually need to create a shared pointer (please note that you would best avoid it since it is more expensive than using the global pointer directly!!)

@KratosMultiphysics/altair @pooyan-dadvand @RiccardoRossi I checked and all changes needed are trivial from our side. Only error is when doing:

~~~cpp
Element::Pointer pcurrent_element= fathers[0]->shared_from_this(); 
~~~

How should be this now?

How should be this now?

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

**Comment by RiccardoRossi** [comment]
@ddiezrod 

sorry, it is probably

~~~
Element::Pointer pcurrent_element= (*fathers[0]).shared_from_this(); 
~~~

**Comment by RiccardoRossi** [comment]
@ddiezrod let me stress that you should really try NOT to create a new pointer if you are not really in need

**Comment by ddiezrod** [comment]
@RiccardoRossi  Thank you very much for the explanation

**Pull request merged**